### PR TITLE
Add the "Decouple the back-end and front-end test through Contract Testing" chapter

### PR DESCRIPTION
**Pull request by NoriSte**
Bring [@NoriSte's article](https://dev.to/noriste/decouple-the-back-end-and-front-end-test-through-contract-testing-112k) to the repo.

### CODE REVIEW

File: sections/real-life-examples/hasura-contract-tests.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
+# Decouple the back-end and front-end test through Contract Testing
+
+<br/><br/>
+
+### One Paragraph Explainer
+
+E2E tests do not scale well for hig applications, with thousands of user flows. Things get even worse if the UI E2E tests are used to test the server too because the result is hundred of slow and coupled tests. In Hasura, an internal working group proposed to decouple the back-end and front-end tests.
```
Comment (muratkeremozcan) [inline sections/real-life-examples/hasura-contract-tests.md:7]: typo "hig applications"

File: sections/real-life-examples/hasura-contract-tests.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+<br/><br/>
+
+## Context
+
+I (a Front-end engineer, keep it in mind) recently championed an initiative aimed at reducing Hasura's CI pipeline duration and the dependence on not-scaling-well E2E tests.
+
+The problem is that **the Server team defensively uses the Hasura Console's** (the main UI of the company) **E2E tests** to check for regressions. If the Console works as expected, the Server works as expected as a by-product.
```
Comment (muratkeremozcan) [inline sections/real-life-examples/hasura-contract-tests.md:15]: lol... so why don't they have api e2e tests?

**Review by muratkeremozcan — COMMENTED** [review]
I don't fully get the layman's takeaway from this article.

At a high level, this sounds like we are running some e2e tests, they are recroding fixtures (which in this context are called contracts), and future tests are made into UI-integration tests. I this correct?

E2e recording to fixutres is a decent idea. What can someone not in Hasura find out more about doing something similar?

How about alternatives? Why doesn't the company use API e2e tests, instead of UI e2e, or UI integration. 

Have you considered Pact, what was the final thoughts. 

I kind of wish to see this kind of analysis at the end, and recommendation for a layman to take some ideas away. Otherwise it's a good story, but does not help the reader so much. The reader needs some practical advice.

**Review by muratkeremozcan — COMMENTED** [review]
I don't fully get the layman's takeaway from this article.

At a high level, this sounds like we are running some e2e tests, they are recroding fixtures (which in this context are called contracts), and future tests are made into UI-integration tests. I this correct?

E2e recording to fixutres is a decent idea. What can someone not in Hasura find out more about doing something similar?

How about alternatives? Why doesn't the company use API e2e tests, instead of UI e2e, or UI integration. 

Can you say more about why Pact failed, what was the final thoughts? UI - to - service approach isn't great, but wasn't the service to service traditional approach a possibility? Must UI be involved?

I kind of wish to see this kind of analysis at the end, and recommendation for a layman to take some ideas away. What should they do in a similar situation. The reader wishes to have some practical advice in a similar situation.

**Comment by NoriSte** [comment]
> I don't fully get the layman's takeaway from this article.
> 
> At a high level, this sounds like we are running some e2e tests, they are recroding fixtures (which in this context are called contracts), and future tests are made into UI-integration tests. Is this correct?
> 
> E2e recording to fixutres is a decent idea. Where can someone not in Hasura find out more about doing something similar?
> 
> How about alternatives? Why doesn't the company use API e2e tests, instead of UI e2e, or UI integration?
> 
> Have you considered Pact? What did you find out, why isn't it a good fit?
> 
> I kind of wish to see this kind of analysis at the end, and recommendation for a layman to take some ideas away. Otherwise it's a good story, but does not help the reader as much as I wish it did.

I agree, let me close this PR and come back on this topic in the future with more examples and suggestions, if any 😊

**Pull request closed without merging**